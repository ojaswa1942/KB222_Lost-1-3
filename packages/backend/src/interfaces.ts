export interface JWTPayload {
  email: string;
  id: string;
}
