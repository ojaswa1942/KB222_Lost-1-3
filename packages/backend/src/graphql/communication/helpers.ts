import { getRepository } from 'typeorm';
import { Room } from '../../database/entity/Room';
import { Channel } from '../../database/entity/Channel';
import { User } from '../../database/entity/User';

export const createRoom = async (name: string, channel: Channel, users: User[]): Promise<Room> => {
  const roomRepo = getRepository(Room);
  const newRoom = roomRepo.create({ name, channel, users });

  return roomRepo.save(newRoom);
};
