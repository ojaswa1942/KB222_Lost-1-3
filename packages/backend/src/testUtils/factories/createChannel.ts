import { Channel } from '../../database/entity/Channel';
import { createEntity } from './createEntity';

export async function createChannel(attributes: Partial<Channel> = {}): Promise<Channel> {
  return createEntity(Channel, {
    ...attributes,
  });
}
