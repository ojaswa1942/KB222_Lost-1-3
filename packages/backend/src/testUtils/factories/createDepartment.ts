import faker from 'faker';

import { Department } from '../../database/entity';
import { createEntity } from './createEntity';

export async function createDepartment(attributes: Partial<Department> = {}): Promise<Department> {
  return createEntity(Department, {
    name: `${faker.name.jobArea()} Department`,
    ...attributes,
  });
}
