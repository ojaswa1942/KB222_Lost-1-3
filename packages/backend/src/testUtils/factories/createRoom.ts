import faker from 'faker';

import { Room } from '../../database/entity/Room';
import { createEntity } from './createEntity';

export async function createRoom(attributes: Partial<Room> = {}): Promise<Room> {
  return createEntity(Room, {
    name: `${faker.address.city()}`,
    ...attributes,
  });
}
