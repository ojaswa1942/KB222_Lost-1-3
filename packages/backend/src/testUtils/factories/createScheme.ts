import faker from 'faker';

import { Scheme } from '../../database/entity';
import { createEntity } from './createEntity';

export async function createScheme(attributes: Partial<Scheme> = {}): Promise<Scheme> {
  return createEntity(Scheme, {
    name: `${faker.name.jobArea()} Scheme`,
    ...attributes,
  });
}
