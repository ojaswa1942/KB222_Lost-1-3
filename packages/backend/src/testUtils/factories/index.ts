export { createUser } from './createUser';
export { createDepartment } from './createDepartment';
