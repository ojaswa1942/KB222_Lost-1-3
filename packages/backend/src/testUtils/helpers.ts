import { ApolloServer } from 'apollo-server-express';
import { ApolloServerTestClient, createTestClient as createApolloTestClient } from 'apollo-server-testing';

import { Context } from '../context';
import { rootSchema } from '../graphql/rootSchema';
import { buildUserLoader } from '../loaders';

export const createTestClient = (ctx?: Partial<Context>): ApolloServerTestClient => {
  const server = new ApolloServer({
    schema: rootSchema,
    context: {
      headers: {},
      req: {},
      res: {
        cookie: () => {
          // do nothing
        },
      },
      userLoader: buildUserLoader(),
      isValid: false,
      ...ctx,
    },
  });

  return createApolloTestClient(server);
};
