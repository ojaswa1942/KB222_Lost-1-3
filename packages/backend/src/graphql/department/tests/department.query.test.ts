import { gql } from 'apollo-server-express';

import { createUser, createDepartment } from '../../../testUtils/factories';
import { createTestClient } from '../../../testUtils/helpers';
import { UserType } from '../../../interfaces';

describe('department query', () => {
  const DepartmentQuery = gql`
    query($input: DepartmentInput!) {
      department(input: $input) {
        id
        name
        users {
          id
        }
        channels {
          id
        }
      }
    }
  `;

  test('fetch department with root', async () => {
    // Given
    const user = await createUser({ email: 'root@example.com', password: 'test', type: UserType.Root });
    const dep = await createDepartment();

    // When
    const res = await createTestClient({
      isValid: true,
      jwt: { id: user.id, email: user.email, type: user.type },
    }).query({
      query: DepartmentQuery,
      variables: {
        input: {
          id: dep.id,
        },
      },
    });

    // Then
    expect(res.errors).toBe(undefined);
    expect(res.data).not.toBe(null);
    expect(res.data?.department).toMatchSnapshot();
  });

  test('fetch department with member user', async () => {
    // Given
    const user = await createUser({ email: 'user@example.com', password: 'test' });
    // const dep = await createDepartment({ departmentRoles: [{ role: DeptRoles.ADMIN, user }] });

    // When
    const res = await createTestClient({
      isValid: true,
      jwt: { id: user.id, email: user.email, type: user.type },
    }).query({
      query: DepartmentQuery,
      variables: {
        input: {
          id: 1,
        },
      },
    });

    // Then
    expect(res.errors).toBe(undefined);
    expect(res.data).not.toBe(null);
    expect(res.data?.department).toMatchSnapshot();
  });

  test('fetch department with non-member user', async () => {
    // Given
    const user = await createUser({ email: 'user@example.com', password: 'test' });
    const dep = await createDepartment();

    // When
    const res = await createTestClient({
      isValid: true,
      jwt: { id: user.id, email: user.email, type: user.type },
    }).query({
      query: DepartmentQuery,
      variables: {
        input: {
          id: dep.id,
        },
      },
    });

    // Then
    expect(res.errors).not.toBe(undefined);

    const error = res.errors[0];
    expect(error.extensions.code).toBe('FORBIDDEN');
  });
});
