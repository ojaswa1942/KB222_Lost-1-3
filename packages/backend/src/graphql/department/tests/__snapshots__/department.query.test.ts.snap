// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`department query fetch department with member user 1`] = `
Object {
  "id": 1,
}
`;

exports[`department query fetch department with root 1`] = `
Object {
  "id": 1,
}
`;
